### PR TITLE
[CDAP-18966] Run kubeJobCleaner only in appfabric pod

### PR DESCRIPTION
**Pull request by kapadiap**
Currently the kubeJobCleaner background thread runs in both AppFabric and Preview services where only 1 is required.

Testing Done:
1. Change cdapmasters CR with a. trace log level for io.cdap.cdap.k8s.runtime b. reduce cleanup interval with program.container.cleaner.interval.mins: 10
2. Run pipeline with runtime param system.runtime.cleanup.disabled=true in k8s native env.
3. observe the appFabric and preview pod logs and confirm cleanup is only running in appfabric.
4. verify the pipeline specific entities are cleaned up after jobcleaner executes.

### CODE REVIEW

File: cdap-kubernetes/src/main/java/io/cdap/cdap/master/environment/k8s/KubeMasterEnvironment.java [hunk 1]
```diff
@@ -331,20 +332,22 @@ public void initialize(MasterEnvironmentContext context) throws IOException, Ill
...
-    if (twillRunnables != null && podLabels != null) {
+    if (twillRunnables != null) {
       for (String twillRunnable : twillRunnables.split(",")) {
         if (podLabels.containsKey(DEFAULT_CONTAINER_LABEL) &&
           podLabels.get(DEFAULT_CONTAINER_LABEL).equals(twillRunnable)) {
           enableMonitor = false;
         }
       }
     }
+    boolean enableKubeJobCleaner = podLabels.containsKey(APPFABRIC_POD_LABEL);
```
Comment (chtyim) [inline cdap-kubernetes/src/main/java/io/cdap/cdap/master/environment/k8s/KubeMasterEnvironment.java]: This is kind of ugly. Can we use the `MasterEnvironment.getTask()` and `MasterEnvironmentTask` instead? That is only executed in app-fabric.

Comment (kapadiap) [reply cdap-kubernetes/src/main/java/io/cdap/cdap/master/environment/k8s/KubeMasterEnvironment.java]: Updated the change to use MasterEnvironmentTask instead.

File: cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/runtime/KubeTwillRunnerService.java [hunk 2]
```diff
@@ -353,6 +340,13 @@ public void onNamespaceDeletion(NamespaceDetail namespaceDetail) throws Exceptio
     }
   }
 
+  /**
+   * Returns the k8s label selector which selects all runs started by the k8s twill runner that has the run id label.
+   */
+  public String getK8sTwillResourcesSelector() {
```
Comment (chtyim) [inline cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/runtime/KubeTwillRunnerService.java]: This method name is repeating this class. This is already a k8s twill runner, so simply call this `getSelector()` is good enough.

File: cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/runtime/KubeTwillRunnerService.java [hunk 3]
```diff
@@ -153,8 +152,6 @@ public class KubeTwillRunnerService implements TwillRunnerService, NamespaceList
   private final Map<String, Map<Type, AppResourceWatcherThread<?>>> resourceWatchers;
   private final Map<String, KubeLiveInfo> liveInfos;
   private final Lock liveInfoLock;
-  private final int jobCleanupIntervalMins;
-  private final int jobCleanBatchSize;
   private final String selector;
   private final boolean enableMonitor;
   private final Gson gson;
```
Comment (chtyim) [inline cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/runtime/KubeTwillRunnerService.java]: I notice that this field is unused. Do we need it?

File: cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/runtime/KubeTwillRunnerService.java [hunk 4]
```diff
@@ -353,6 +340,13 @@ public void onNamespaceDeletion(NamespaceDetail namespaceDetail) throws Exceptio
     }
   }
 
+  /**
+   * Returns the k8s label selector which selects all runs started by the k8s twill runner that has the run id label.
+   */
+  public String getK8sTwillResourcesSelector() {
+    return this.selector;
```
Comment (chtyim) [inline cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/runtime/KubeTwillRunnerService.java]: Remove `this.`

File: cdap-kubernetes/src/main/java/io/cdap/cdap/master/environment/k8s/KubeJobCleaner.java [hunk 5]
```diff
@@ -14,41 +14,47 @@
...
   private final String selector;
   private final int batchSize;
+  private final long delayMillis;
 
-  KubeJobCleaner(BatchV1Api batchV1Api, String selector, int batchSize) {
-    this.batchV1Api = batchV1Api;
+  KubeJobCleaner(String selector, int batchSize, long delayMin) {
     this.selector = selector;
     this.batchSize = batchSize;
+    this.delayMillis = delayMin * 60 * 1000;
```
Comment (chtyim) [inline cdap-kubernetes/src/main/java/io/cdap/cdap/master/environment/k8s/KubeJobCleaner.java]: Use `TimeUnit.MINUTES.toMillis` instead.

File: cdap-kubernetes/src/main/java/io/cdap/cdap/master/environment/k8s/KubeJobCleaner.java [hunk 6]
```diff
@@ -82,6 +88,9 @@ public void run() {
           }
         }
         continuationToken = jobs.getMetadata().getContinue();
+      } catch (IOException e) {
+        LOG.warn("Failed to cleanup jobs. Error creating batch api client.", e);
+        break;
```
Comment (chtyim) [inline cdap-kubernetes/src/main/java/io/cdap/cdap/master/environment/k8s/KubeJobCleaner.java]: Should we try on this error as well? There could be temporary disconnection to the API server?

File: cdap-master-spi/src/main/java/io/cdap/cdap/master/spi/environment/MasterEnvironment.java [hunk 7]
```diff
@@ -53,11 +54,11 @@ default void destroy() {
   }
 
   /**
-   * Returns a {@link Optional} {@link MasterEnvironmentTask} to be executed periodically.
-   * It is guaranteed that there is no concurrent call to the task returned.
+   * Returns a list of {@link MasterEnvironmentTask}s to be executed periodically.
+   * It is guaranteed that there is no concurrent call to each task returned.
```
Comment (chtyim) [inline cdap-master-spi/src/main/java/io/cdap/cdap/master/spi/environment/MasterEnvironment.java]: I don't think the current implementation is not guaranteeing no concurrent task execution?

File: cdap-kubernetes/src/main/java/io/cdap/cdap/master/environment/k8s/KubeMasterEnvironment.java [hunk 8]
```diff
@@ -388,8 +389,8 @@ public Supplier<TwillRunnerService> getTwillRunnerSupplier() {
   }
 
   @Override
-  public Optional<MasterEnvironmentTask> getTask() {
-    return Optional.ofNullable(podKillerTask);
+  public List<MasterEnvironmentTask> getTasks() {
```
Comment (chtyim) [inline cdap-kubernetes/src/main/java/io/cdap/cdap/master/environment/k8s/KubeMasterEnvironment.java]: I think using `Collection` is better as ordering is not important.

File: cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/runtime/KubeTwillRunnerService.java (hunk 3 above)
Comment (kapadiap) [reply cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/runtime/KubeTwillRunnerService.java]: removed this unused field

File: cdap-kubernetes/src/main/java/io/cdap/cdap/master/environment/k8s/KubeJobCleaner.java (hunk 6 above)
Comment (kapadiap) [reply cdap-kubernetes/src/main/java/io/cdap/cdap/master/environment/k8s/KubeJobCleaner.java]: hmm, I thought this was only looking at kubeconfig files cached locally, but it does reach out to the server as a last resort. We can retry on this error.

File: cdap-master-spi/src/main/java/io/cdap/cdap/master/spi/environment/MasterEnvironment.java (hunk 7 above)
Comment (kapadiap) [reply cdap-master-spi/src/main/java/io/cdap/cdap/master/spi/environment/MasterEnvironment.java]: I am not too sure myself. The underlying AbstractSchedulerService provides a guarantee that the run iterations of a specific task will never run concurrently. However we can create multiple instances of the same task and schedule them all to run concurrently.

Comment (kapadiap) [reply cdap-master-spi/src/main/java/io/cdap/cdap/master/spi/environment/MasterEnvironment.java]: Let me just remove the guarantee claim.

File: cdap-kubernetes/src/main/java/io/cdap/cdap/master/environment/k8s/KubeJobCleaner.java [hunk 9]
```diff
@@ -14,41 +14,47 @@
...
 import java.util.concurrent.TimeUnit;
 
 /**
  * Kubernetes Job cleaner that scans and deletes completed jobs across all namespaces.
  */
-class KubeJobCleaner implements Runnable {
+class KubeJobCleaner implements MasterEnvironmentTask {
   private static final Logger LOG = LoggerFactory.getLogger(KubeJobCleaner.class);
-  private final BatchV1Api batchV1Api;
+  private volatile BatchV1Api batchV1Api;
```
Comment (masoud-io) [inline cdap-kubernetes/src/main/java/io/cdap/cdap/master/environment/k8s/KubeJobCleaner.java:40]: Add comments why this needs to be volatile

**Review by masoud-io — COMMENTED** [review]
Please ensure that jobcleaner service does not run in system worker pods.

File: cdap-kubernetes/src/main/java/io/cdap/cdap/master/environment/k8s/KubeMasterEnvironment.java [hunk 10]
```diff
@@ -219,19 +218,16 @@ public class KubeMasterEnvironment implements MasterEnvironment {
...
   private String workloadIdentityPool;
   private long workloadIdentityServiceAccountTokenTTLSeconds;
   private String programCpuMultiplier;
   private String cdapInstallNamespace;
+  private KubeJobCleaner kubeJobCleaner;
 
-  public KubeMasterEnvironment() {
-    gson = new Gson();
-  }
+  public KubeMasterEnvironment() {}
```
Comment (chtyim) [inline cdap-kubernetes/src/main/java/io/cdap/cdap/master/environment/k8s/KubeMasterEnvironment.java]: We can remove this too since Java always generate a no-op default constructor if the class doesn't have a constructor.

File: cdap-master/src/main/java/io/cdap/cdap/master/environment/k8s/AppFabricServiceMain.java [hunk 11]
```diff
@@ -164,8 +164,8 @@ protected void addServices(Injector injector, List<? super Service> services,
       services.add(injector.getInstance(SystemWorkerServiceLauncher.class));
     }
 
-    // Optionally adds the master environment task
-    masterEnv.getTask().ifPresent(task -> services.add(new MasterTaskExecutorService(task, masterEnvContext)));
+    // Adds the master environment tasks
+    masterEnv.getTasks().forEach(task -> services.add(new MasterTaskExecutorService(task, masterEnvContext)));
```
Comment (chtyim) [inline cdap-master/src/main/java/io/cdap/cdap/master/environment/k8s/AppFabricServiceMain.java:168]: The MasterTaskExecutorService has a hardcode thread name. Since now we can run more than one master env task, it's better to have the thread name contains the task name as well. The `MasterEnvironmentTask` can have a getName method that default to the simple class name (`getClass().getSimpleName()`)

File: cdap-kubernetes/src/main/java/io/cdap/cdap/master/environment/k8s/KubeMasterEnvironment.java [hunk 12]
```diff
@@ -388,8 +384,8 @@ public Supplier<TwillRunnerService> getTwillRunnerSupplier() {
   }
 
   @Override
-  public Optional<MasterEnvironmentTask> getTask() {
-    return Optional.ofNullable(podKillerTask);
+  public Collection<MasterEnvironmentTask> getTasks() {
+    return SetUtils.hashSet(podKillerTask, kubeJobCleaner);
```
Comment (chtyim) [inline cdap-kubernetes/src/main/java/io/cdap/cdap/master/environment/k8s/KubeMasterEnvironment.java]: No need to use apache common lib in this case. Just return `Arrays.asList(podKillerTask, kubeJobCleaner)`.

File: cdap-kubernetes/src/main/java/io/cdap/cdap/master/environment/k8s/KubeMasterEnvironment.java [hunk 13]
```diff
@@ -342,14 +337,15 @@ public void initialize(MasterEnvironmentContext context) throws IOException, Ill
...
-                                             Integer.parseInt(conf.getOrDefault(JOB_CLEANUP_INTERVAL, "60")),
-                                             Integer.parseInt(conf.getOrDefault(JOB_CLEANUP_BATCH_SIZE, "1000")),
                                              enableMonitor,
                                              workloadIdentityEnabled,
                                              workloadLauncherRoleNameForNamespace,
                                              workloadLauncherRoleNameForCluster,
                                              workloadIdentityPool,
                                              workloadIdentityProvider);
+    kubeJobCleaner = new KubeJobCleaner(twillRunner.getSelector(),
+                                        Integer.parseInt(conf.getOrDefault(JOB_CLEANUP_BATCH_SIZE, "1000")),
```
Comment (chtyim) [inline cdap-kubernetes/src/main/java/io/cdap/cdap/master/environment/k8s/KubeMasterEnvironment.java]: Please define the default as private constant.

File: cdap-master/src/main/java/io/cdap/cdap/master/environment/k8s/MasterTaskExecutorService.java [hunk 14]
```diff
@@ -47,7 +47,8 @@ final class MasterTaskExecutorService extends AbstractScheduledService {
 
   @Override
   protected ScheduledExecutorService executor() {
-    executor = Executors.newSingleThreadScheduledExecutor(Threads.createDaemonThreadFactory("master-env-executor"));
+    executor = Executors.newSingleThreadScheduledExecutor(Threads.createDaemonThreadFactory(
+      "master-env-task-" + task.getName() + "-%d"));
```
Comment (chtyim) [inline cdap-master/src/main/java/io/cdap/cdap/master/environment/k8s/MasterTaskExecutorService.java]: Since this is a single thread executor, you don't need the `-%d` part.

Comment (kapadiap) [reply cdap-master/src/main/java/io/cdap/cdap/master/environment/k8s/MasterTaskExecutorService.java]: I see, I added it because the underlying ThreadFactory always expects such postfix integer format. I will remove this, the extra arg would be safely ignored.

**Comment by kapadiap** [comment]
> Please ensure that jobcleaner service does not run in system worker pods.

Looked at the logs, the system worker pods are not running jobcleaner. The system worker pods use the MasterEnvironmentMain rather than the AppFabricServiceMain and hence do not schedule the kube cleaner task.

**Review by chtyim — APPROVED** [review]

File: cdap-kubernetes/src/main/java/io/cdap/cdap/master/environment/k8s/KubeMasterEnvironment.java [hunk 15]
```diff
@@ -397,7 +397,12 @@ public Supplier<TwillRunnerService> getTwillRunnerSupplier() {
 
   @Override
   public Collection<MasterEnvironmentTask> getTasks() {
-    return Arrays.asList(podKillerTask, kubeJobCleaner);
+    Set<MasterEnvironmentTask> set = new HashSet<>(2);
```
Comment (chtyim) [inline cdap-kubernetes/src/main/java/io/cdap/cdap/master/environment/k8s/KubeMasterEnvironment.java:400]: Remove the `2`.

File: cdap-kubernetes/src/main/java/io/cdap/cdap/master/environment/k8s/KubeMasterEnvironment.java [hunk 16]
```diff
@@ -397,7 +397,12 @@ public Supplier<TwillRunnerService> getTwillRunnerSupplier() {
...
+    set.add(kubeJobCleaner);
+    if (podKillerTask != null) {
```
Comment (chtyim) [inline cdap-kubernetes/src/main/java/io/cdap/cdap/master/environment/k8s/KubeMasterEnvironment.java:402]: Why is this needed? A task shouldn't be null?

Comment (kapadiap) [reply cdap-kubernetes/src/main/java/io/cdap/cdap/master/environment/k8s/KubeMasterEnvironment.java:402]: The podKillerTask is an optional and only initialized if some configurations are explicitly provided.
